File: computer-science/data-structures/heaps/PriorityQueue.py
class PriorityQueue(object):
    def __init__(self, heap = [], size = 0):
        self.heap = heap
        self.size = size

    def swim(self, k):

        while (k > 1 and self.isLess(int(k / 2))):
            self.exchange(int(k / 2), k)
            k = int(k / 2)

    def sink(self, k):
        while (2 * k <= self.size):
            j = k * 2

            if (j < self.size and self.isLess(j, j + 1)): j = j + 1
            if (self.isLess(k, j) is False): break

            self.exchange(k, j)
            k = j

    def insert(self, v):
        self.size = self.size + 1

        self.heap[self.size] = v
        self.sink(self.size)

    def removeMax(self):
        m = self.heap[1]
        self.size = self.size - 1

        self.exchange(1, self.size)
        self.heap[self.size + 1] = None
        self.sink(1)

    def size(self):
        return self.size

    def isEmpty(self):
        return self.size is 0

    def isLess(self, x, y):
        return self.heap[x] < self.heap[y]

    def exchange(self, x, y):
        t = self.heap[x]
        self.heap[x] = self.heap[y]
        self.heap[y] = t

File: computer-science/data-structures/heaps/PriorityQueue.js
/**
 * JavaScript Custom Implementation of a Priority Queue data structure
 *
 * michael@softwareontheshore.com
 */

 'use strict';

function PriorityQueue() {
    this.heap = [];
    this.N = 0;
}

PriorityQueue.prototype.size = function() {
    return this.N;
}

PriorityQueue.prototype.isEmpty = function() {
    return this.N === 0;
}

PriorityQueue.prototype.insert = function(v) {
    this.heap[++this.N] = v;
    this.swim(this.N);
}

PriorityQueue.prototype.removeMax = function() {
    const max = this.heap[1];

    this.exchange(1, this.N--);
    this.heap[this.N + 1] = null;

    this.sink(1);

    return max;
}

/* restore the heap order when a new node is added at the bottom or a node's priority
* has increased (bottom-up restore)
*/
PriorityQueue.prototype.swim = function(k) {

    /* compare child node to parent and exchange if necessary */
    while (k > 1 && this.isLess(k / 2, k)) {
        this.exchange(k / 2, k);
    }
}

/* restore the heap order when a node's priority has decreased, such as when
 * the root is exchanged with a smaller key (top-down restore)
 */
PriorityQueue.prototype.sink = function(k) {

    /* compare parent node with two children and exchange if necessary */
    while (2 * k <= this.N) {
        const j = 2 * k;

        if (j < this.N && this.isLess(j, j + 1)) { j++; }
        if (!this.isLess(k, j)) { break; }

        this.exchange(k, j);
        k = j;
    }
}

/* compare two nodes to determine if x is less than y */
PriorityQueue.prototype.isLess = function(x, y) {
    return this.heap[x] < this.heap[y];
}

/* exchange the values of two nodes to restore heap order */
PriorityQueue.prototype.exchange = function(x, y) {
    const node = this.heap[x];

    this.heap[x] = this.heap[y];
    this.heap[y] = node;
}

const pq = new PriorityQueue();

pq.insert(10);
pq.insert(11);
pq.insert(12);
pq.insert(13);

console.log(pq.heap);
